tfconfig/load_hcl.go

```diff
@@ -60,6 +60,20 @@ func loadModule(dir string) (*Module, Diagnostics) {
 							if _, exists := mod.RequiredProviders[name]; !exists {
 								mod.RequiredProviders[name] = req
 							} else {
+								if req.Source != "" {
+									source := mod.RequiredProviders[name].Source
+									if source != "" && source != req.Source {
+										diags = append(diags, &hcl.Diagnostic{
+											Severity: hcl.DiagWarning,
```

Comment:
Looks great! Only one concern: I think this should be an error, not a warning. What do you think?

Reply:
I agree, but I can't convince it to work. If I use a `DiagError` severity, the output for this test skips the required providers section altogether:

```
--- FAIL: TestRenderMarkdown (0.01s)
    --- FAIL: TestRenderMarkdown/provider-source-invalid (0.00s)
        markdown_test.go:48:
            # Module `testdata/provider-source-invalid`

             !=
            # Module `testdata/provider-source-invalid`

            Provider Requirements:
            * **bat (`abc/bat`):** `1.0.0`
            * **foo (`abc/foo`):** `~> 2.0.0, ~> 2.1.0`

            ## Problems

            ## Warning: Multiple provider source attriubtes

            (at `testdata/provider-source-invalid/provider-source-invalid.tf` line 15)

            Found multiple source attributes for provider bat: "abc/bat", "baz/bat"
```

I believe this is because of this fallback code if we generate an error with our loader: https://github.com/hashicorp/terraform-config-inspect/blob/36d6f40eb7d7627b5ebdc25fe8aca21343dd82af/tfconfig/load.go#L24-L31

Since the legacy parser doesn't generate an error, we end up skipping the required providers block altogether 😕

I'm not sure how to address that. Do you have any ideas?

Reply:
I'm not sure I follow - the test output you've posted looks correct to me; could you write the test that expects an error and specifically check for that error? 

`Found multiple source attributes for provider bat: "abc/bat", "baz/bat"`

Reply:
Sorry, I wasn't clear (and neither is the test output). This is from me running the new test on this branch ([files here](https://github.com/hashicorp/terraform-config-inspect/tree/e8736b60c9e844b14f549e570373f97c1aca8177/tfconfig/testdata/provider-source-invalid)), which fails when I change the severity to error. The failure mode is an empty markdown file, rather than `"warning" != "error"` as I'd hoped.

That is, the output is:

```markdown
# Module `testdata/provider-source-invalid`
```

instead of:

```markdown
# Module `testdata/provider-source-invalid`

Provider Requirements:
* **bat (`abc/bat`):** `1.0.0`
* **foo (`abc/foo`):** `~> 2.0.0, ~> 2.1.0`

## Problems

## Error: Multiple provider source attriubtes

(at `testdata/provider-source-invalid/provider-source-invalid.tf` line 15)

Found multiple source attributes for provider bat: "abc/bat", "baz/bat"
```

Reply:
I think what you're seeing here is an unfortunate interaction between the two different parsing modes in this library...

`terraform-config-inspect` first parses with the HCL 2-based parser. If it gets an error, it then tries to parse with the HCL 1-based parser. If the latter succeeds then its result is returned. If the latter fails then the whole module is considered to be in error and we return the errors from _the HCL 2-based parser_ because we expect them to be better.

The behavior you're reporting above looks to me like the HCL 1-based parser is succeeding, presumably because it doesn't know about the `required_providers` block (that was a Terraform 0.12 feature) and is just ignoring it.

To make this fail in the way you want, I think it might be necessary to add an extra check to the HCL 1-based parser to force it to always fail if a `required_providers` block is present, thus ensuring it can never "win" and suppress the errors returned by the HCL 2-based parser. That's gross, but I think it will work, if I'm right about the root cause.

Reply:
The test cases in `TestRenderMarkdown` (historically) are only testing valid modules, and other tests check for expected errors on config load. 

In looking at your branch I'd say that this is correct: the markdown should fail to render (ie, there should be no valid `*.out.json` to even test against) since we expect an error in that situation.

Reply:
@apparentlymart Thanks! I added the extra check to the legacy parser in b19c2c0 and now we have an error instead of a warning 🎉 

@mildwonkey I _think_ this might be the right approach now…? We do still render a Markdown file, but it includes the error diagnostic, so hopefully that's the desired behaviour.

Reply:
I have concerns about returning a possibly-inaccurate result (in markdown) alongside the error. I would expect no output, or at least empty `"required_providers"` output, in this case. 

I'm not sure how this tool is getting used, so I could be way off on my risk assessment, but my fear is seeing inaccurate data in the registry, or causing issue with other third-party tools that assume if they get a result, the config was valid, and therefor they do not check for diagnostic errors.

Having said this, it looks great, and if that's consistent with how `terraform_config_inspect` renders markdown in the face of _other_ errors I won't make a fuss!

Reply:
I think it's consistent with existing behaviour to still render a Markdown file (and a JSON file). [Here's an example of existing Markdown test output](https://github.com/hashicorp/terraform-config-inspect/blob/36d6f40eb7d7627b5ebdc25fe8aca21343dd82af/tfconfig/testdata/type-errors/type-errors.out.md), and [corresponding JSON which includes nodes for invalid objects](https://github.com/hashicorp/terraform-config-inspect/blob/36d6f40eb7d7627b5ebdc25fe8aca21343dd82af/tfconfig/testdata/type-errors/type-errors.out.json).

But I'm also not sure how the tool is used, and if we need to change that behaviour then I'd be happy to do so.

Reply:
I don't know why I didn't think of this yesterday, but I generally ask @justincampbell to review PRs to this codebase since it's used in the registry. If he doesn't have any concerns about this change that's good enough for us!